DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,795 @@
+manner has the advantage that no parser modifications are required, thus leaving the language grammar
+unchanged.
+
+```d
+import std.stdio;
+
+struct A
+{
+    this(ref A rhs) { writeln("x"); }     // copy constructor
+}
+
+void main()
+{
+    A a;
+    A b = a;        // calls copy constructor implicitly
+    A c = A(b);     // calls constructor explicitly
+}
+```
+
+The copy constructor may also be called explicitly since it is also a constructor.
+
+The argument to the copy constructor is passed by reference in order to avoid infinite recursion (passing by
+value would require a copy of the `struct`, which would be made by calling the copy constructor, thus leading
+to an infinite chain of calls).
+
+Type qualifiers may be applied to the parameter of the copy constructor, but also to the function itself,
+in order to facilitate the ability to describe copies between objects of different mutability levels. The type
+qualifiers are optional.
+
+### Semantics
+
+This section discusses all the aspects of the semantic analysis and interaction between the copy
+constructor and other components.
+
+#### Copy constructor and postblit cohabitation
+
+In order to assure a smooth transition from postblit to copy constructor, the following
+strategy is employed: if a `struct` defines a postblit (user-defined or generated) all
+copy constructor definitions will be ignored for that particular `struct` and the postblit
+will be preferred. Existing code bases that do not use the postblit may start using the
+copy constructor without any problems, while codebases that rely on the postblit may
+start writing new code using the copy constructor and remove the deprecated postblit
+from their code.
+
+The transition from postblit to copy constructor may be simply achieved by replacing the postblit
+declaration `this(this)` with `this(ref inout(typeof(this)) rhs) inout`.
+
+#### Copy constructor usage
+
+The copy constructor is implicitly used by the compiler whenever a `struct` variable is initialized
+as a copy of another variable of the same unqualified type:
+
+1. When a variable is explicitly initialized:
+
+```d
+struct A
+{
+    this(ref A rhs) {}
+}
+
+void main()
+{
+    A a;
+    A b = a; // copy constructor gets called
+    b = a;   // assignment, not initialization
+}
+```
+
+2. When a parameter is passed by value to a function:
+
+```d
+struct A
+{
+    this(ref A another) {}
+}
+
+void fun(A a) {}
+
+void main()
+{
+    A a;
+    fun(a);    // copy constructor gets called
+}
+```
+
+3. When a parameter is returned by value from a function and NRVO cannot be performed:
+
+```d
+struct A
+{
+    this(ref A another)
+    {
+        writeln("cpCtor");
+    }
+}
+
+A fun()
+{
+    A a;
+    return a;
+}
+
+A a;
+A gun()
+{
+    return a;
+}
+
+void main()
+{
+    A a;
+    a = fun();      // NRVO - no copy constructor call
+    A b;
+    b = gun();      // NRVO cannot be performed, copy constructor is called
+}
+```
```

Comment:
That is true, but the effect is as if the copy constructor would get called on the returned value. I wanted to avoid going into compiler specific generated code such as `(A __tmp; __tmp.copyCtor(a); __tmp)` for the `return a` line in the gun function.